OrbitCore/LinuxPerfUtils.h

```diff
@@ -0,0 +1,99 @@
+//-----------------------------------
+// Copyright Pierric Gimmig 2013-2019
+//-----------------------------------
+#pragma once
+
+#include <asm/unistd.h>
+#include <linux/perf_event.h>
+#include <sys/errno.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+
+// This struct must be in sync with the SAMPLE_TYPE_FLAGS.
+struct perf_sample_id {
+    uint32_t pid, tid;  /* if PERF_SAMPLE_TID set */
+    uint64_t time;      /* if PERF_SAMPLE_TIME set */
+    uint32_t cpu, res;  /* if PERF_SAMPLE_CPU set */
+};
+
+struct perf_event_fork_exit {
+    struct perf_event_header header;
+    uint32_t    pid, ppid;
+    uint32_t    tid, ptid;
+    uint64_t    time;
+    struct perf_sample_id sample_id;
+};
+
+struct perf_event_lost {
+    struct perf_event_header header;
+    uint64_t    id;
+    uint64_t    lost;
+    struct perf_sample_id sample_id;
+};
+
+// TODO: This struct might change. We should read this from debugfs.
+struct sched_switch_tp {
+    uint16_t common_type;
+    unsigned char common_flags;
+	unsigned char common_preempt_count;
+	int32_t common_pid;
+
+	char prev_comm[16];
+	int32_t prev_pid;
+	int32_t prev_prio;
+	int64_t prev_state;
+	char next_comm[16];
+	int32_t next_pid;
+	int32_t next_prio;
+};
+
+
+// This struct must be in sync with the SAMPLE_TYPE_FLAGS.
+// TODO: rename T
+template<typename RecordData>
+struct __attribute__((__packed__)) perf_event_record {
+    struct perf_event_header header;
+    uint32_t                 pid, tid;   /* if PERF_SAMPLE_TID */
+    uint64_t                 time;       /* if PERF_SAMPLE_TIME */
+    uint32_t                 cpu, res;   /* if PERF_SAMPLE_CPU */
+    uint32_t                 size;       /* if PERF_SAMPLE_RAW */
+    RecordData               data;       /* if PERF_SAMPLE_RAW */
+};
+
+//-----------------------------------------------------------------------------
+namespace LinuxPerfUtils
+{
+    // TODO: these constants should be made available in some settings
+    static const uint32_t STACK_SIZE = (1u << 16u) - 8; // TODO: the stack is not yet recorded
```

Comment:
Does the TODO mean that the stack size here is incorrect until the stack has been recorded? Maybe make this is a constant whose meaning is actually always correct and then record the actual stack size in a variable that always contains the correct value.

Reply:
The stack size is actually correct, but it would simply not used in perf_event_open internally.